// @holywater-tech/ads-builder/framework/components/character/CharacterAnimationConfig.js
export const STATES = {
    STATIC: 'static',
    ANGER: 'anger',
    ENVY: 'envy',
    SHAME: 'shame',
    FRIGHT: 'fright',
    JOY: 'joy',
    ASTONISHMENT: 'astonishment',
    COLD: 'cold',
    HOT: 'hot',
    IDLE: 'idle',
    MEDITATION: 'meditation',
    EMBRACE: 'embrace',
};

export const EMOTIONS = {
    DEFAULT: 'face_default',
    FLIRTY: 'face_flirty',
    SAD: 'face_sad',
    JOY: 'face_joy',
    SURPRISE: 'face_surprise',
    SHY: 'face_shy',
    ANGRY: 'face_angry',
};

export const BODY_PARTS = {
    BODY: 'body',
    HAIR_BACK: 'hair_back',
    FACE: 'face',
    FACE_RED: 'face_red',
    SWEAT: 'sweat',
    HAIR_FRONT: 'hair_front',
    DRESS: 'dress',
    ARM_LEFT: 'arm_left',
    FOREARM_LEFT: 'forearm_left',
    HAND_LEFT: 'hand_left',
    ARM_RIGHT: 'arm_right',
    FOREARM_RIGHT: 'forearm_right',
    HAND_RIGHT: 'hand_right',
};

export class CharacterAnimationConfig {
    constructor() {
        this.body = {};

        this.body[BODY_PARTS.ARM_LEFT] = { img: 'arm_left', originX: 0.61, originY: 0.315, depth: 1 };
        this.body[BODY_PARTS.BODY] = { img: 'body', originX: 0.5, originY: 0.5, depth: 2 };
        this.body[BODY_PARTS.HAIR_BACK] = { img: 'hair_back', originX: 0.47, originY: 0.25, depth: 3 };
        this.body[BODY_PARTS.FACE] = { img: 'face_default', originX: 0.47, originY: 0.25, depth: 4 };
        this.body[BODY_PARTS.FACE_RED] = { img: 'face_red', originX: 0.47, originY: 0.25, depth: 5, alpha: 0 };
        this.body[BODY_PARTS.DRESS] = { img: 'dress', originX: 0.5, originY: 0.5, depth: 6 };
        this.body[BODY_PARTS.ARM_RIGHT] = { img: 'arm_right', originX: 0.355, originY: 0.325, depth: 7 };
        this.body[BODY_PARTS.HAIR_FRONT] = { img: 'hair_front', originX: 0.47, originY: 0.25, depth: 8 };
        this.body[BODY_PARTS.FOREARM_LEFT] = { img: 'forearm_left', originX: 0.673, originY: 0.492, depth: 9 };
        this.body[BODY_PARTS.HAND_LEFT] = { img: 'hand_left', originX: 0.78, originY: 0.63, depth: 10 };
        this.body[BODY_PARTS.FOREARM_RIGHT] = { img: 'forearm_right', originX: 0.32, originY: 0.51, depth: 11 };
        this.body[BODY_PARTS.HAND_RIGHT] = { img: 'hand_right', originX: 0.335, originY: 0.655, depth: 12 };
        this.body[BODY_PARTS.SWEAT] = { img: 'sweat', originX: 0.47, originY: 0.25, depth: 13, alpha: 0 };
    }
}
