// @holywater-tech/ads-builder/framework/components/ui/Item.js
import { EVENTS_DEFAULT } from '../EventsDispatcher';
import Utils from '../../Utils';

/* mode could be glow(default value) or fill */
export default class Item extends Phaser.GameObjects.Container {
    constructor(scene, item, isOnce = true, mode = 'glow') {
        super(scene, 0, 0);
        this.tweens = scene.tweens;
        this.img = item;
        this.item = this.scene.add.image(0, 0, 'atlas', `${item}_item`).setDepth(4);
        this.glow = this.scene.add
            .image(0, 0, 'atlas', `item_box_${mode}`)
            .setDepth(mode === 'glow' ? 1 : 3)
            .setAlpha(0);
        this.base = this.scene.add.image(0, 0, 'atlas', 'item_box').setDepth(2);
        this.add([this.item, this.glow, this.base]);
        this._sort();
        
        this.base.setInteractive()[isOnce ? 'once' : 'on']('pointerdown', this.onClick, this);

        this.scene.emitter.on(EVENTS_DEFAULT.ON_ITEM_CLICK, this.onItemClick, this);
        this.on('destroy', this.onDestroy, this);
    }

    onDestroy() {
        this.scene.emitter.off(EVENTS_DEFAULT.ON_ITEM_CLICK, this.onItemClick, this);
    }

    disable() {
        this.base.disableInteractive();
    }

    addGlow() {
        this.scene.pj.fadeIn(this.glow);
        return this;
    }

    onClick() {
        Utils.addAudio(this.scene, 'tap', 0.5, false);
        this.isSelected = true;
        this.addGlow();
        this.scene.emitter.emit(EVENTS_DEFAULT.ON_ITEM_CLICK, this);
    }

    onItemClick(obj) {
        if (obj === this) {
            return;
        }
        this.isSelected = false;
        this.scene.pj.fadeOut(this.glow);
    }
}
