// @holywater-tech/ads-builder/framework/components/ui/PlayButton.js
import { EVENTS_DEFAULT } from '../EventsDispatcher';

export default class PlayButton extends Phaser.GameObjects.Container {
    constructor(scene, sprite = 'play_now_btn', glow = false) {
        super(scene, 0, 0);
        this.tweens = scene.tweens;
        this.setCustomAlign('Bottom').addProperties(['pos', 'scale']).setAlpha(0);

        this.button = this.scene.add.image(0, 0, sprite).setDepth(2);
        this.glow = this.scene.add.image(0, 0, `${sprite}_glow`).setDepth(1).setScale(0.9).setVisible(glow);

        this.add([this.button, this.glow]);
        this._sort();
        this.addPosition();
        this.addAnimation();
        this.scene.pj.fadeIn(this);
    }

    addPosition() {
        this.setCustomPosition(0, -200, 0, -150).setCustomScale(1, 1, 1, 1);
    }

    addAnimation() {
        const onClick = () => {
            this.scene.emitter.emit(EVENTS_DEFAULT.OPEN_STORE);
        };
        this.tweens.add({
            targets: this.glow.visible ? this.glow : this.button,
            scale: '*=1.1',
            duration: 500,
            yoyo: true,
            repeat: -1,
        });
        this.button.setInteractive().on('pointerdown', onClick, this);
    }

    _sort() {
        this.sort('depth');
    }
}
