// @holywater-tech/ads-builder/framework/components/ui/PressButton.js
import { EVENTS_DEFAULT } from '../EventsDispatcher';

export default class PressButton extends Phaser.GameObjects.Container {
    constructor(scene, sprite, isDisabled, emitMessage) {
        super(scene, 0, 0);

        this.setCustomAlign('Bottom').setCustomPosition(0, -150).setAlpha(0).setEmitMessage(emitMessage);
        this.scene.pj.fadeIn(this);

        this.disabled = this.scene.add
            .image(0, 0, 'atlas', `${sprite}_disabled`)
            .setDepth(1)
            .setInteractive()
            .setAlpha(isDisabled ? 1 : 0)
            .once('pointerdown', this.onDisabledClick, this);
        this.enabled = this.scene.add
            .image(0, 0, 'atlas', `${sprite}_enabled`)
            .setDepth(2)
            .setAlpha(isDisabled ? 0 : 1)
            .setInteractive()
            .once('pointerdown', this.onClick, this);
        this.pressed = this.scene.add.image(0, 0, 'atlas', `${sprite}_pressed`).setDepth(3).setAlpha(0);
        this.add([this.disabled, this.enabled, this.pressed]);
        this._sort();
    }

    setEmitMessage(msg) {
        this.emitMessage = msg;
        return this;
    }

    enable() {
        this.scene.pj.fadeOut(this.disabled).fadeIn(this.enabled);
    }

    onDisabledClick() {
        this.scene.pj.shake(this);
    }

    onClick() {
        const onComplete = () => {
            this.scene.emitter.emit(this.emitMessage || EVENTS_DEFAULT.BUTTON_PRESSED);
        };

        this.scene.tweens.add({ targets: this.pressed, alpha: 1, yoyo: true, duration: 250 });
        this.scene.tweens.add({ targets: this, scale: '*=0.9', duration: 250, yoyo: true, onComplete });
    }
}
